File: packages/flutter-im-ui/lib/src/models/interaction_state.dart
import 'message_lifecycle.dart';
import '../components/flare_transfer_progress.dart';

enum FlareComposerInteractionMode {
  idle,
  typing,
  mentioning,
  replying,
  editing,
  uploading,
  recording,
  sending,
  sendBlocked,
  offline,
  readOnly,
  slowMode,
  permissionDenied,
}

enum FlareComposerInteractionAction {
  type,
  send,
  newline,
  attach,
  mention,
  reply,
  edit,
  record,
  cancel,
  retry,
}

class FlareComposerInteractionInput {
  const FlareComposerInteractionInput({
    this.hasText = false,
    this.mentioning = false,
    this.replying = false,
    this.editing = false,
    this.uploading = false,
    this.recording = false,
    this.sending = false,
    this.sendBlocked = false,
    this.online = true,
    this.readOnly = false,
    this.slowMode = false,
    this.permissionGranted = true,
  });
  final bool hasText, mentioning, replying, editing, uploading, recording;
  final bool sending,
      sendBlocked,
      online,
      readOnly,
      slowMode,
      permissionGranted;
}

FlareComposerInteractionMode resolveComposerMode(
  FlareComposerInteractionInput input,
) {
  if (input.readOnly) return FlareComposerInteractionMode.readOnly;
  if (!input.permissionGranted)
    return FlareComposerInteractionMode.permissionDenied;
  if (!input.online) return FlareComposerInteractionMode.offline;
  if (input.slowMode) return FlareComposerInteractionMode.slowMode;
  if (input.sendBlocked) return FlareComposerInteractionMode.sendBlocked;
  if (input.sending) return FlareComposerInteractionMode.sending;
  if (input.recording) return FlareComposerInteractionMode.recording;
  if (input.uploading) return FlareComposerInteractionMode.uploading;
  if (input.editing) return FlareComposerInteractionMode.editing;
  if (input.mentioning) return FlareComposerInteractionMode.mentioning;
  if (input.replying) return FlareComposerInteractionMode.replying;
  return input.hasText
      ? FlareComposerInteractionMode.typing
      : FlareComposerInteractionMode.idle;
}

Set<FlareComposerInteractionAction> composerAllowedActions(
  FlareComposerInteractionMode mode,
) {
  const base = {
    FlareComposerInteractionAction.type,
    FlareComposerInteractionAction.send,
    FlareComposerInteractionAction.newline,
    FlareComposerInteractionAction.attach,
    FlareComposerInteractionAction.mention,
    FlareComposerInteractionAction.reply,
    FlareComposerInteractionAction.edit,
    FlareComposerInteractionAction.record,
  };
  return switch (mode) {
    FlareComposerInteractionMode.readOnly => const {},
    FlareComposerInteractionMode.permissionDenied => const {
      FlareComposerInteractionAction.cancel,
    },
    FlareComposerInteractionMode.offline => const {
      FlareComposerInteractionAction.type,
      FlareComposerInteractionAction.newline,
      FlareComposerInteractionAction.attach,
      FlareComposerInteractionAction.cancel,
      FlareComposerInteractionAction.retry,
    },
    FlareComposerInteractionMode.slowMode ||
    FlareComposerInteractionMode.sendBlocked => const {
      FlareComposerInteractionAction.type,
      FlareComposerInteractionAction.newline,
      FlareComposerInteractionAction.attach,
      FlareComposerInteractionAction.mention,
      FlareComposerInteractionAction.cancel,
    },
    FlareComposerInteractionMode.sending => const {
      FlareComposerInteractionAction.cancel,
    },
    FlareComposerInteractionMode.recording => const {
      FlareComposerInteractionAction.send,
      FlareComposerInteractionAction.cancel,
    },
    FlareComposerInteractionMode.uploading => const {
      FlareComposerInteractionAction.type,
      FlareComposerInteractionAction.newline,
      FlareComposerInteractionAction.attach,
      FlareComposerInteractionAction.cancel,
    },
    FlareComposerInteractionMode.editing ||
    FlareComposerInteractionMode.mentioning ||
    FlareComposerInteractionMode.replying => const {
      FlareComposerInteractionAction.type,
      FlareComposerInteractionAction.newline,
      FlareComposerInteractionAction.send,
      FlareComposerInteractionAction.attach,
      FlareComposerInteractionAction.mention,
      FlareComposerInteractionAction.cancel,
    },
    _ => base,
  };
}

enum FlareMessageInteractionMode {
  normal,
  hover,
  selected,
  multiSelected,
  sending,
  sent,
  delivered,
  read,
  failed,
  retrying,
  edited,
  recalled,
  deleted,
  ephemeral,
  expired,
}

FlareMessageInteractionMode resolveMessageMode(
  FlareMessageLifecycle lifecycle, {
  bool hovered = false,
  bool selected = false,
  bool multiSelect = false,
  bool retrying = false,
}) {
  if (lifecycle.mutation == FlareMessageMutationState.deleted)
    return FlareMessageInteractionMode.deleted;
  if (lifecycle.ephemeral == FlareMessageEphemeralState.expired)
    return FlareMessageInteractionMode.expired;
  if (lifecycle.mutation == FlareMessageMutationState.recalled)
    return FlareMessageInteractionMode.recalled;
  if (retrying) return FlareMessageInteractionMode.retrying;
  if (lifecycle.send == FlareMessageSendState.failed ||
      lifecycle.transfer == FlareTransferState.failed)
    return FlareMessageInteractionMode.failed;
  if (lifecycle.ephemeral != FlareMessageEphemeralState.none)
    return FlareMessageInteractionMode.ephemeral;
  if (lifecycle.mutation == FlareMessageMutationState.edited)
    return FlareMessageInteractionMode.edited;
  if (multiSelect && selected) return FlareMessageInteractionMode.multiSelected;
  if (selected) return FlareMessageInteractionMode.selected;
  if (hovered) return FlareMessageInteractionMode.hover;
  if (lifecycle.send == FlareMessageSendState.sending ||
      lifecycle.send == FlareMessageSendState.draft)
    return FlareMessageInteractionMode.sending;
  if (lifecycle.read == FlareMessageReadState.read)
    return FlareMessageInteractionMode.read;
  if (lifecycle.delivery != FlareMessageLifecycleDeliveryState.serverAccepted)
    return FlareMessageInteractionMode.delivered;
  return lifecycle.send == FlareMessageSendState.sent
      ? FlareMessageInteractionMode.sent
      : FlareMessageInteractionMode.normal;
}

enum FlareMessageCapability {
  reply,
  reaction,
  copy,
  forward,
  mergeForward,
  multiSelect,
  edit,
  delete,
  recall,
  pin,
  unpin,
  save,
  translate,
  report,
  retry,
  openThread,
  jumpToQuote,
}

class FlareMessageCapabilityInput {
  const FlareMessageCapabilityInput({
    required this.lifecycle,
    required this.own,
    this.canModerate = false,
    this.supportsCopy = true,
    this.supportsSave = false,
    this.supportsTranslate = false,
    this.supportsMergeForward = false,
    this.reportable = true,
    this.pinned = false,
    this.hasThread = false,
    this.hasQuote = false,
  });
  final FlareMessageLifecycle lifecycle;
  final bool own, canModerate, supportsCopy, supportsSave, supportsTranslate;
  final bool supportsMergeForward, reportable, pinned, hasThread, hasQuote;
}

Set<FlareMessageCapability> resolveMessageCapabilities(
  FlareMessageCapabilityInput input,
) {
  final lifecycle = input.lifecycle;
  final terminal =
      lifecycle.mutation == FlareMessageMutationState.deleted ||
      lifecycle.mutation == FlareMessageMutationState.recalled ||
      lifecycle.ephemeral == FlareMessageEphemeralState.expired;
  final failed =
      lifecycle.send == FlareMessageSendState.failed ||
      lifecycle.transfer == FlareTransferState.failed;
  if (terminal) return input.canModerate ? {FlareMessageCapability.delete} : {};
  final result = <FlareMessageCapability>{FlareMessageCapability.multiSelect};
  if (input.own || input.canModerate) result.add(FlareMessageCapability.delete);
  if (failed) {
    if (input.own) result.add(FlareMessageCapability.retry);
  } else {
    result.addAll({
      FlareMessageCapability.reply,
      FlareMessageCapability.reaction,
      FlareMessageCapability.forward,
    });
    result.add(
      input.pinned ? FlareMessageCapability.unpin : FlareMessageCapability.pin,
    );
    if (input.supportsMergeForward)
      result.add(FlareMessageCapability.mergeForward);
    if (input.hasThread) result.add(FlareMessageCapability.openThread);
    if (input.hasQuote) result.add(FlareMessageCapability.jumpToQuote);
    if (input.supportsCopy) result.add(FlareMessageCapability.copy);
    if (input.supportsSave) result.add(FlareMessageCapability.save);
    if (input.supportsTranslate) result.add(FlareMessageCapability.translate);
    if (input.own && lifecycle.send == FlareMessageSendState.sent) {
      result.addAll({
        FlareMessageCapability.edit,
        FlareMessageCapability.recall,
      });
    }
    if (!input.own && input.reportable)
      result.add(FlareMessageCapability.report);
  }
  return result;
}

enum FlareMessageActionGroup { primary, organize, message, destructive }

enum FlareMessageActionPresentation {
  contextMenu,
  hoverToolbar,
  commandPalette,
  actionSheet,
  bottomSheet,
}

class FlareMessageAction {
  const FlareMessageAction({
    required this.id,
    required this.group,
    required this.destructive,
    required this.promoted,
  });
  final FlareMessageCapability id;
  final FlareMessageActionGroup group;
  final bool destructive, promoted;
}

List<FlareMessageAction> resolveMessageActions(
  FlareMessageCapabilityInput input,
  FlareMessageActionPresentation presentation,
) {
  const primary = {
    FlareMessageCapability.reply,
    FlareMessageCapability.reaction,
    FlareMessageCapability.copy,
    FlareMessageCapability.forward,
    FlareMessageCapability.openThread,
  };
  const organize = {
    FlareMessageCapability.pin,
    FlareMessageCapability.unpin,
    FlareMessageCapability.save,
    FlareMessageCapability.translate,
    FlareMessageCapability.multiSelect,
    FlareMessageCapability.mergeForward,
  };
  const destructive = {
    FlareMessageCapability.recall,
    FlareMessageCapability.delete,
    FlareMessageCapability.report,
  };
  return resolveMessageCapabilities(input)
      .map(
        (id) => FlareMessageAction(
          id: id,
          group: destructive.contains(id)
              ? FlareMessageActionGroup.destructive
              : organize.contains(id)
              ? FlareMessageActionGroup.organize
              : primary.contains(id)
              ? FlareMessageActionGroup.primary
              : FlareMessageActionGroup.message,
          destructive: destructive.contains(id),
          promoted:
              presentation == FlareMessageActionPresentation.hoverToolbar &&
              primary.contains(id),
        ),
      )
      .toList(growable: false);
}

class FlareSelectionState {
  const FlareSelectionState({this.selectedIds = const {}, this.anchorId});
  final Set<String> selectedIds;
  final String? anchorId;
}

sealed class FlareSelectionEvent {
  const FlareSelectionEvent();
}

class FlareSelectionToggle extends FlareSelectionEvent {
  const FlareSelectionToggle(this.id);
  final String id;
}

class FlareSelectionReplace extends FlareSelectionEvent {
  const FlareSelectionReplace(this.id);
  final String id;
}

class FlareSelectionExtend extends FlareSelectionEvent {
  const FlareSelectionExtend(this.id, this.orderedIds);
  final String id;
  final List<String> orderedIds;
}

class FlareSelectionClear extends FlareSelectionEvent {
  const FlareSelectionClear();
}

FlareSelectionState reduceSelection(
  FlareSelectionState state,
  FlareSelectionEvent event,
) {
  if (event is FlareSelectionClear) return const FlareSelectionState();
  if (event is FlareSelectionReplace)
    return FlareSelectionState(selectedIds: {event.id}, anchorId: event.id);
  if (event is FlareSelectionToggle) {
    final next = {...state.selectedIds};
    next.contains(event.id) ? next.remove(event.id) : next.add(event.id);
    return FlareSelectionState(selectedIds: next, anchorId: event.id);
  }
  final extend = event as FlareSelectionExtend;
  final anchor = state.anchorId == null
      ? -1
      : extend.orderedIds.indexOf(state.anchorId!);
  final target = extend.orderedIds.indexOf(extend.id);
  if (anchor < 0 || target < 0)
    return FlareSelectionState(selectedIds: {extend.id}, anchorId: extend.id);
  final start = anchor < target ? anchor : target;
  final end = anchor > target ? anchor : target;
  return FlareSelectionState(
    selectedIds: extend.orderedIds.sublist(start, end + 1).toSet(),
    anchorId: state.anchorId,
  );
}

enum FlareDesktopShortcutAction {
  search,
  commandPalette,
  newConversation,
  focusComposer,
  send,
  newline,
  closeOverlay,
  previousConversation,
  nextConversation,
  toggleDetails,
  reply,
  edit,
  delete,
  copy,
  forward,
  openThread,
}

enum FlareDesktopShortcutScope { global, composer, conversation }

FlareDesktopShortcutAction? resolveDesktopShortcut(
  String key, {
  bool primary = false,
  bool alt = false,
  bool shift = false,
  FlareDesktopShortcutScope scope = FlareDesktopShortcutScope.global,
}) {
  final normalized = key.toLowerCase();
  if (normalized == 'escape') return FlareDesktopShortcutAction.closeOverlay;
  if (primary && normalized == 'k')
    return FlareDesktopShortcutAction.commandPalette;
  if (primary && normalized == 'f') return FlareDesktopShortcutAction.search;
  if (primary && normalized == 'n')
    return FlareDesktopShortcutAction.newConversation;
  if (primary && shift && normalized == 'd')
    return FlareDesktopShortcutAction.toggleDetails;
  if (alt && normalized == 'arrowup')
    return FlareDesktopShortcutAction.previousConversation;
  if (alt && normalized == 'arrowdown')
    return FlareDesktopShortcutAction.nextConversation;
  if (scope == FlareDesktopShortcutScope.composer && normalized == 'enter')
    return primary
        ? FlareDesktopShortcutAction.send
        : FlareDesktopShortcutAction.newline;
  if (scope == FlareDesktopShortcutScope.conversation &&
      primary &&
      normalized == 'c')
    return FlareDesktopShortcutAction.copy;
  if (scope == FlareDesktopShortcutScope.conversation &&
      !primary &&
      !alt &&
      normalized == 'r')
    return FlareDesktopShortcutAction.reply;
  if (scope == FlareDesktopShortcutScope.conversation &&
      !primary &&
      !alt &&
      normalized == 'e')
    return FlareDesktopShortcutAction.edit;
  if (scope == FlareDesktopShortcutScope.conversation &&
      !primary &&
      !alt &&
      normalized == 'f')
    return FlareDesktopShortcutAction.forward;
  if (scope == FlareDesktopShortcutScope.conversation &&
      !primary &&
      !alt &&
      normalized == 't')
    return FlareDesktopShortcutAction.openThread;
  if (scope == FlareDesktopShortcutScope.conversation &&
      (normalized == 'delete' || normalized == 'backspace'))
    return FlareDesktopShortcutAction.delete;
  return null;
}

enum FlareSwipeIntent { none, reply, conversationLeading, conversationTrailing }

FlareSwipeIntent resolveSwipeIntent({
  required double deltaX,
  required double deltaY,
  required bool message,
  bool rtl = false,
  bool multiSelect = false,
  double threshold = 56,
}) {
  if (multiSelect ||
      deltaX.abs() < threshold ||
      deltaX.abs() < deltaY.abs() * 1.25)
    return FlareSwipeIntent.none;
  final leading = rtl ? deltaX < 0 : deltaX > 0;
  if (message) return leading ? FlareSwipeIntent.reply : FlareSwipeIntent.none;
  return leading
      ? FlareSwipeIntent.conversationLeading
      : FlareSwipeIntent.conversationTrailing;
}

File: packages/android-im-ui/src/main/kotlin/com/flare/im/ui/InteractionState.kt
package com.flare.im.ui

enum class FlareComposerInteractionMode {
    Idle, Typing, Mentioning, Replying, Editing, Uploading, Recording, Sending,
    SendBlocked, Offline, ReadOnly, SlowMode, PermissionDenied,
}

enum class FlareComposerInteractionAction { Type, Send, Newline, Attach, Mention, Reply, Edit, Record, Cancel, Retry }

data class FlareComposerInteractionInput(
    val hasText: Boolean = false,
    val mentioning: Boolean = false,
    val replying: Boolean = false,
    val editing: Boolean = false,
    val uploading: Boolean = false,
    val recording: Boolean = false,
    val sending: Boolean = false,
    val sendBlocked: Boolean = false,
    val online: Boolean = true,
    val readOnly: Boolean = false,
    val slowMode: Boolean = false,
    val permissionGranted: Boolean = true,
)

fun resolveComposerMode(input: FlareComposerInteractionInput): FlareComposerInteractionMode = when {
    input.readOnly -> FlareComposerInteractionMode.ReadOnly
    !input.permissionGranted -> FlareComposerInteractionMode.PermissionDenied
    !input.online -> FlareComposerInteractionMode.Offline
    input.slowMode -> FlareComposerInteractionMode.SlowMode
    input.sendBlocked -> FlareComposerInteractionMode.SendBlocked
    input.sending -> FlareComposerInteractionMode.Sending
    input.recording -> FlareComposerInteractionMode.Recording
    input.uploading -> FlareComposerInteractionMode.Uploading
    input.editing -> FlareComposerInteractionMode.Editing
    input.mentioning -> FlareComposerInteractionMode.Mentioning
    input.replying -> FlareComposerInteractionMode.Replying
    input.hasText -> FlareComposerInteractionMode.Typing
    else -> FlareComposerInteractionMode.Idle
}

fun composerAllowedActions(mode: FlareComposerInteractionMode): Set<FlareComposerInteractionAction> = when (mode) {
    FlareComposerInteractionMode.ReadOnly -> emptySet()
    FlareComposerInteractionMode.PermissionDenied -> setOf(FlareComposerInteractionAction.Cancel)
    FlareComposerInteractionMode.Offline -> setOf(FlareComposerInteractionAction.Type, FlareComposerInteractionAction.Newline, FlareComposerInteractionAction.Attach, FlareComposerInteractionAction.Cancel, FlareComposerInteractionAction.Retry)
    FlareComposerInteractionMode.SlowMode, FlareComposerInteractionMode.SendBlocked -> setOf(FlareComposerInteractionAction.Type, FlareComposerInteractionAction.Newline, FlareComposerInteractionAction.Attach, FlareComposerInteractionAction.Mention, FlareComposerInteractionAction.Cancel)
    FlareComposerInteractionMode.Sending -> setOf(FlareComposerInteractionAction.Cancel)
    FlareComposerInteractionMode.Recording -> setOf(FlareComposerInteractionAction.Send, FlareComposerInteractionAction.Cancel)
    FlareComposerInteractionMode.Uploading -> setOf(FlareComposerInteractionAction.Type, FlareComposerInteractionAction.Newline, FlareComposerInteractionAction.Attach, FlareComposerInteractionAction.Cancel)
    FlareComposerInteractionMode.Editing, FlareComposerInteractionMode.Mentioning, FlareComposerInteractionMode.Replying -> setOf(FlareComposerInteractionAction.Type, FlareComposerInteractionAction.Newline, FlareComposerInteractionAction.Send, FlareComposerInteractionAction.Attach, FlareComposerInteractionAction.Mention, FlareComposerInteractionAction.Cancel)
    else -> setOf(FlareComposerInteractionAction.Type, FlareComposerInteractionAction.Send, FlareComposerInteractionAction.Newline, FlareComposerInteractionAction.Attach, FlareComposerInteractionAction.Mention, FlareComposerInteractionAction.Reply, FlareComposerInteractionAction.Edit, FlareComposerInteractionAction.Record)
}

enum class FlareMessageInteractionMode { Normal, Hover, Selected, MultiSelected, Sending, Sent, Delivered, Read, Failed, Retrying, Edited, Recalled, Deleted, Ephemeral, Expired }

fun resolveMessageMode(
    lifecycle: FlareMessageLifecycle,
    hovered: Boolean = false,
    selected: Boolean = false,
    multiSelect: Boolean = false,
    retrying: Boolean = false,
): FlareMessageInteractionMode = when {
    lifecycle.mutation == FlareMessageMutationState.Deleted -> FlareMessageInteractionMode.Deleted
    lifecycle.ephemeral == FlareMessageEphemeralState.Expired -> FlareMessageInteractionMode.Expired
    lifecycle.mutation == FlareMessageMutationState.Recalled -> FlareMessageInteractionMode.Recalled
    retrying -> FlareMessageInteractionMode.Retrying
    lifecycle.send == FlareMessageSendState.Failed || lifecycle.transfer == FlareTransferState.Failed -> FlareMessageInteractionMode.Failed
    lifecycle.ephemeral != FlareMessageEphemeralState.None -> FlareMessageInteractionMode.Ephemeral
    lifecycle.mutation == FlareMessageMutationState.Edited -> FlareMessageInteractionMode.Edited
    multiSelect && selected -> FlareMessageInteractionMode.MultiSelected
    selected -> FlareMessageInteractionMode.Selected
    hovered -> FlareMessageInteractionMode.Hover
    lifecycle.send == FlareMessageSendState.Sending || lifecycle.send == FlareMessageSendState.Draft -> FlareMessageInteractionMode.Sending
    lifecycle.read == FlareMessageReadState.Read -> FlareMessageInteractionMode.Read
    lifecycle.delivery != FlareMessageLifecycleDeliveryState.ServerAccepted -> FlareMessageInteractionMode.Delivered
    lifecycle.send == FlareMessageSendState.Sent -> FlareMessageInteractionMode.Sent
    else -> FlareMessageInteractionMode.Normal
}

enum class FlareMessageCapability {
    Reply, Reaction, Copy, Forward, MergeForward, MultiSelect, Edit, Delete, Recall,
    Pin, Unpin, Save, Translate, Report, Retry, OpenThread, JumpToQuote,
}

data class FlareMessageCapabilityInput(
    val lifecycle: FlareMessageLifecycle,
    val own: Boolean,
    val canModerate: Boolean = false,
    val supportsCopy: Boolean = true,
    val supportsSave: Boolean = false,
    val supportsTranslate: Boolean = false,
    val supportsMergeForward: Boolean = false,
    val reportable: Boolean = true,
    val pinned: Boolean = false,
    val hasThread: Boolean = false,
    val hasQuote: Boolean = false,
)

fun resolveMessageCapabilities(input: FlareMessageCapabilityInput): Set<FlareMessageCapability> {
    val lifecycle = input.lifecycle
    val terminal = lifecycle.mutation == FlareMessageMutationState.Deleted ||
        lifecycle.mutation == FlareMessageMutationState.Recalled ||
        lifecycle.ephemeral == FlareMessageEphemeralState.Expired
    val failed = lifecycle.send == FlareMessageSendState.Failed || lifecycle.transfer == FlareTransferState.Failed
    if (terminal) return if (input.canModerate) setOf(FlareMessageCapability.Delete) else emptySet()
    return buildSet {
        add(FlareMessageCapability.MultiSelect)
        if (input.own || input.canModerate) add(FlareMessageCapability.Delete)
        if (failed) {
            if (input.own) add(FlareMessageCapability.Retry)
        } else {
            addAll(setOf(FlareMessageCapability.Reply, FlareMessageCapability.Reaction, FlareMessageCapability.Forward))
            add(if (input.pinned) FlareMessageCapability.Unpin else FlareMessageCapability.Pin)
            if (input.supportsMergeForward) add(FlareMessageCapability.MergeForward)
            if (input.hasThread) add(FlareMessageCapability.OpenThread)
            if (input.hasQuote) add(FlareMessageCapability.JumpToQuote)
            if (input.supportsCopy) add(FlareMessageCapability.Copy)
            if (input.supportsSave) add(FlareMessageCapability.Save)
            if (input.supportsTranslate) add(FlareMessageCapability.Translate)
            if (input.own && lifecycle.send == FlareMessageSendState.Sent) addAll(setOf(FlareMessageCapability.Edit, FlareMessageCapability.Recall))
            if (!input.own && input.reportable) add(FlareMessageCapability.Report)
        }
    }
}

enum class FlareMessageActionGroup { Primary, Organize, Message, Destructive }
enum class FlareMessageActionPresentation { ContextMenu, HoverToolbar, CommandPalette, ActionSheet, BottomSheet }
data class FlareMessageAction(val id: FlareMessageCapability, val group: FlareMessageActionGroup, val destructive: Boolean, val promoted: Boolean)

fun resolveMessageActions(input: FlareMessageCapabilityInput, presentation: FlareMessageActionPresentation): List<FlareMessageAction> {
    val primary = setOf(FlareMessageCapability.Reply, FlareMessageCapability.Reaction, FlareMessageCapability.Copy, FlareMessageCapability.Forward, FlareMessageCapability.OpenThread)
    val organize = setOf(FlareMessageCapability.Pin, FlareMessageCapability.Unpin, FlareMessageCapability.Save, FlareMessageCapability.Translate, FlareMessageCapability.MultiSelect, FlareMessageCapability.MergeForward)
    val destructive = setOf(FlareMessageCapability.Recall, FlareMessageCapability.Delete, FlareMessageCapability.Report)
    return resolveMessageCapabilities(input).map { id ->
        FlareMessageAction(
            id,
            when (id) { in destructive -> FlareMessageActionGroup.Destructive; in organize -> FlareMessageActionGroup.Organize; in primary -> FlareMessageActionGroup.Primary; else -> FlareMessageActionGroup.Message },
            id in destructive,
            presentation == FlareMessageActionPresentation.HoverToolbar && id in primary,
        )
    }
}

data class FlareSelectionState(val selectedIds: Set<String> = emptySet(), val anchorId: String? = null)
sealed interface FlareSelectionEvent
data class FlareSelectionToggle(val id: String) : FlareSelectionEvent
data class FlareSelectionReplace(val id: String) : FlareSelectionEvent
data class FlareSelectionExtend(val id: String, val orderedIds: List<String>) : FlareSelectionEvent
data object FlareSelectionClear : FlareSelectionEvent

fun reduceSelection(state: FlareSelectionState, event: FlareSelectionEvent): FlareSelectionState = when (event) {
    FlareSelectionClear -> FlareSelectionState()
    is FlareSelectionReplace -> FlareSelectionState(setOf(event.id), event.id)
    is FlareSelectionToggle -> FlareSelectionState(state.selectedIds.toMutableSet().apply { if (!add(event.id)) remove(event.id) }, event.id)
    is FlareSelectionExtend -> {
        val anchor = state.anchorId?.let(event.orderedIds::indexOf) ?: -1
        val target = event.orderedIds.indexOf(event.id)
        if (anchor < 0 || target < 0) FlareSelectionState(setOf(event.id), event.id)
        else FlareSelectionState(event.orderedIds.subList(minOf(anchor, target), maxOf(anchor, target) + 1).toSet(), state.anchorId)
    }
}

enum class FlareDesktopShortcutAction { Search, CommandPalette, NewConversation, FocusComposer, Send, Newline, CloseOverlay, PreviousConversation, NextConversation, ToggleDetails, Reply, Edit, Delete, Copy, Forward, OpenThread }
enum class FlareDesktopShortcutScope { Global, Composer, Conversation }

fun resolveDesktopShortcut(key: String, primary: Boolean = false, alt: Boolean = false, shift: Boolean = false, scope: FlareDesktopShortcutScope = FlareDesktopShortcutScope.Global): FlareDesktopShortcutAction? {
    val normalized = key.lowercase()
    return when {
        normalized == "escape" -> FlareDesktopShortcutAction.CloseOverlay
        primary && normalized == "k" -> FlareDesktopShortcutAction.CommandPalette
        primary && normalized == "f" -> FlareDesktopShortcutAction.Search
        primary && normalized == "n" -> FlareDesktopShortcutAction.NewConversation
        primary && shift && normalized == "d" -> FlareDesktopShortcutAction.ToggleDetails
        alt && normalized == "arrowup" -> FlareDesktopShortcutAction.PreviousConversation
        alt && normalized == "arrowdown" -> FlareDesktopShortcutAction.NextConversation
        scope == FlareDesktopShortcutScope.Composer && normalized == "enter" -> if (primary) FlareDesktopShortcutAction.Send else FlareDesktopShortcutAction.Newline
        scope == FlareDesktopShortcutScope.Conversation && primary && normalized == "c" -> FlareDesktopShortcutAction.Copy
        scope == FlareDesktopShortcutScope.Conversation && !primary && !alt && normalized == "r" -> FlareDesktopShortcutAction.Reply
        scope == FlareDesktopShortcutScope.Conversation && !primary && !alt && normalized == "e" -> FlareDesktopShortcutAction.Edit
        scope == FlareDesktopShortcutScope.Conversation && !primary && !alt && normalized == "f" -> FlareDesktopShortcutAction.Forward
        scope == FlareDesktopShortcutScope.Conversation && !primary && !alt && normalized == "t" -> FlareDesktopShortcutAction.OpenThread
        scope == FlareDesktopShortcutScope.Conversation && normalized in setOf("delete", "backspace") -> FlareDesktopShortcutAction.Delete
        else -> null
    }
}

enum class FlareSwipeIntent { None, Reply, ConversationLeading, ConversationTrailing }

fun resolveSwipeIntent(deltaX: Float, deltaY: Float, message: Boolean, rtl: Boolean = false, multiSelect: Boolean = false, threshold: Float = 56f): FlareSwipeIntent {
    if (multiSelect || kotlin.math.abs(deltaX) < threshold || kotlin.math.abs(deltaX) < kotlin.math.abs(deltaY) * 1.25f) return FlareSwipeIntent.None
    val leading = if (rtl) deltaX < 0 else deltaX > 0
    return if (message) {
        if (leading) FlareSwipeIntent.Reply else FlareSwipeIntent.None
    } else if (leading) FlareSwipeIntent.ConversationLeading else FlareSwipeIntent.ConversationTrailing
}
